prov/cxi/src/cxip_msg_hpc.c

```diff
@@ -3317,6 +3325,186 @@ static int cxip_recv_sw_matcher(struct cxip_rxc_hpc *rxc, struct cxip_req *req,
+	if (ret)
+		return ret;
+
+	/* req->data_len, req->tag, req->data must be set later. req->buf may
+	 * be overwritten later.
+	 */
+	req->context = (uint64_t)context;
+
+	req->flags = FI_RECV | (flags & FI_COMPLETION);
+	if (tagged)
+		req->flags |= FI_TAGGED;
+	else
+		req->flags |= FI_MSG;
+
+	req->recv.cntr = comp_cntr ? comp_cntr : rxc->recv_cntr;
+	req->recv.match_id = match_id;
+	req->recv.tag = tag;
+	req->recv.ignore = ignore;
+	req->recv.flags = flags;
+	req->recv.tagged = tagged;
+	req->recv.multi_recv = (flags & FI_MULTI_RECV ? true : false);
+
+	*req_out = req;
+
+	return FI_SUCCESS;
+
+lock_err:
+	ofi_genlock_lock(&rxc->ep_obj->lock);
+	return ret;
+}
+
+int cxip_unexp_start(struct fi_peer_rx_entry *rx_entry)
+{
+	int ret;
+	struct cxip_ux_send *ux;
+	union cxip_match_bits ux_mb;
+	struct cxip_req *req;
+	struct cxip_rxc *rxc;
+
+	ux = rx_entry->peer_context;
+	ux_mb.raw = ux->put_ev.tgt_long.match_bits;
+	rxc = ux->rxc;
+
+	ret = cxip_recv_req_init(rxc, rx_entry->iov[0].iov_base,
+				rx_entry->iov[0].iov_len, rx_entry->addr,
+				rx_entry->tag, 0, rx_entry->flags,
+				ux_mb.tagged, rx_entry->context, NULL, &req);
+	if (ret)
+		return ret;
+
+	req->rx_entry = rx_entry;
+
+	ret = cxip_recv_sw_matched(req, ux);
+
+	if (ux->req && ux->req->type == CXIP_REQ_RBUF)
+		cxip_req_buf_ux_free(ux);
+	else
+		free(ux);
+
+	RXC_DBG(rxc,
+		"Software match, req: %p ux_send: %p (sw_ux_list_len: %u)\n",
+		req, ux, req->recv.rxc->sw_ux_list_len);
+
+	return ret;
+}
+
+static int cxip_process_srx_ux_matcher(struct cxip_rxc *rxc,
+		struct fid_peer_srx *owner_srx, struct cxip_ux_send *ux)
+{
+	int ret;
+	uint32_t ux_init;
+	union cxip_match_bits ux_mb;
+	struct fi_peer_rx_entry *rx_entry = NULL;
+	struct cxip_req *req;
+	uint16_t vni;
+	struct fi_peer_match_attr match = {0};
+
+	/* stash the rxc because we're going to need it if the peer
+	 * address isn't already inserted into the AV table.
+	 */
+	ux->rxc = rxc;
+	ux_init = ux->put_ev.tgt_long.initiator.initiator.process;
+	vni = ux->put_ev.tgt_long.vni;
+
+	match.addr = cxip_recv_req_src_addr(rxc, ux_init, vni, true);
+
+	ux_mb.raw = ux->put_ev.tgt_long.match_bits;
+
+	if (ux_mb.tagged) {
+		match.tag = ux_mb.tag;
+		ret = owner_srx->owner_ops->get_tag(owner_srx, &match, &rx_entry);
+	} else {
+		ret = owner_srx->owner_ops->get_msg(owner_srx, &match, &rx_entry);
+	}
+
+	/* return it back to the caller */
+	ux->rx_entry = rx_entry;
+
+	if (ret == -FI_ENOENT) {
+		/* this is used when the owner calls start_msg */
+		rx_entry->peer_context = ux;
+		return -FI_ENOMSG;
+	} else if (ret) {
+		return ret;
+	}
+
+	ret = cxip_recv_req_init(rxc, rx_entry->iov[0].iov_base,
+				rx_entry->iov[0].iov_len, rx_entry->addr,
+				rx_entry->tag, 0, rx_entry->flags,
+				ux_mb.tagged, rx_entry->context, NULL, &req);
+	if (ret)
+		return ret;
+
+	req->rx_entry = rx_entry;
+
+	ret = cxip_recv_sw_matched(req, ux);
+
+	if (ux->req && ux->req->type == CXIP_REQ_RBUF)
+		cxip_req_buf_ux_free(ux);
+	else
+		free(ux);
```

Comment:
Is it safe to free this ux without consulting ret?

Reply:
I check for EAGAIN.

Reply:
I may be missing something... line 3497 return code is not being checked before freeing the ux. Looking at cxip_req_buf_progress_pending_ux(), this will result in the now freed ux being passed in again.

Reply:
sorry, I changed it int he cxip_unexp_start() and not here. I'll update the patch momentarily.